File: solutions/typescript/robot-simulator/1/robot-simulator.ts
export class InvalidInputError extends Error {
  constructor(message?: string) {
    super()
    this.message = message || 'Invalid Input'
  }
}

type Direction = 'north' | 'east' | 'south' | 'west'
type Coordinates = [number, number]

export class Robot {
  #bearing: Direction = 'north';
  #coordinates: Coordinates = [0, 0];
  
  get bearing(): Direction {
    return this.#bearing;
  }

  get coordinates(): Coordinates {
    return this.#coordinates;
  }

  turnRight(){
    switch(this.#bearing){
      case 'north':
        this.#bearing = 'east';
        break;
      case 'east':
        this.#bearing = 'south';
        break;
      case 'south':
        this.#bearing = 'west';
        break;
      case 'west':
        this.#bearing = 'north';
        break;
      default:
        throw new Error("Invalid bearing");
    }
  }

  turnLeft(){
    switch(this.#bearing){
      case 'north':
        this.#bearing = 'west';
        break;
      case 'west':
        this.#bearing = 'south';
        break;
      case 'south':
        this.#bearing = 'east';
        break;
      case 'east':
        this.#bearing = 'north';
        break;
      default:
        throw new Error("Invalid bearing");
    }
  }

  advance(){
    switch(this.#bearing){
      case 'north':
        this.#coordinates[1]++;
        break;
      case 'east':
        this.#coordinates[0]++;
        break;
      case 'south':
        this.#coordinates[1]--;
        break;
      case 'west':
        this.#coordinates[0]--;
        break;
      default:
        throw new Error("Invalid bearing");
    }
  }

  place(loc: { x: number; y: number; direction: string }) {
    if (loc.direction) {
      if (loc.direction !== 'north' && loc.direction !== 'east' && loc.direction !== 'south' && loc.direction !== 'west'){
        throw new InvalidInputError();
      }
      
      this.#bearing = loc.direction as Direction;
    } 
    
    if (loc.x !== undefined) this.#coordinates[0] = loc.x;
    if (loc.y !== undefined) this.#coordinates[1] = loc.y;
  }

  evaluate(instructions: string) {
    for (let instruction of instructions.split("")){
      switch(instruction){
        case 'R':
          this.turnRight();
          break;
        case 'L':
          this.turnLeft();
          break;
        case 'A':
          this.advance();
          break;
      }
    }
  }
}

File: solutions/cpp/robot-simulator/1/robot_simulator.cpp
#include "robot_simulator.h"

namespace robot_simulator
{
    Robot::Robot(position_t position, enum Bearing bearing) : position{position}, bearing{bearing} {};
    const position_t Robot::get_position() const
    {
        return position;
    }

    enum Bearing Robot::get_bearing() const
    {
        return bearing;
    };

    void Robot::turn_right()
    {
        switch (bearing)
        {
        case Bearing::NORTH:
            bearing = Bearing::EAST;
            break;
        case Bearing::EAST:
            bearing = Bearing::SOUTH;
            break;
        case Bearing::SOUTH:
            bearing = Bearing::WEST;
            break;
        case Bearing::WEST:
            bearing = Bearing::NORTH;
            break;
        }
    }
    void Robot::turn_left()
    {
        switch (bearing)
        {
        case Bearing::NORTH:
            bearing = Bearing::WEST;
            break;
        case Bearing::WEST:
            bearing = Bearing::SOUTH;
            break;
        case Bearing::SOUTH:
            bearing = Bearing::EAST;
            break;
        case Bearing::EAST:
            bearing = Bearing::NORTH;
            break;
        }
    }

    void Robot::advance()
    {
        switch (bearing)
        {
        case Bearing::NORTH:
            position.second++;
            break;
        case Bearing::WEST:
            position.first--;
            break;
        case Bearing::SOUTH:
            position.second--;
            break;
        case Bearing::EAST:
            position.first++;
            break;
        }
    };

    void Robot::execute_sequence(const std::string &sequence)
    {
        for (char step : sequence)
        {
            switch (step)
            {
            case 'R':
                this->turn_right();
                break;
            case 'L':
                this->turn_left();
                break;
            case 'A':
                this->advance();
                break;
            }
        }
    }

} // namespace robot_simulator
